lib/i18n/backend/lazy_loadable.rb

```
@@ -0,0 +1,160 @@
+      # followed by an "_" demarcation to separate proceeding text.
+      def filenames_for_current_locale
+        I18n.load_path.flatten.select do |path|
+          LocaleExtractor.locale_from_path(path) == I18n.locale &&
+          supported_extension?(path)
+        end
+      end
+
+      # Parse the load path and extract all locales.
+      def available_locales
+        if lazy_load?
+          I18n.load_path.map { |path| LocaleExtractor.locale_from_path(path) }
+        else
+          super
+        end
+      end
+
+      protected
+
+      # Load translations from files that belong to the current locale.
+      def init_translations
+        invalid_files = []
+
+        if lazy_load?
+          load_translations(filenames_for_current_locale) { |file, translations| invalid_files << file unless file_named_correctly?(file, translations) }
+          initialized_locales[I18n.locale] = true
+        else
+          load_translations { |file, translations| invalid_files << file unless file_named_correctly?(file, translations) }
+          @initialized = true
+        end
+
+        raise InvalidFilenames.new(invalid_files) unless invalid_files.empty?
+      end
+
+      def initialized_locales
+        @initialized_locales ||= Hash.new(false)
+      end
+
+      private
+
+      def lazy_load?
+        @lazy_load
+      end
+
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
+
+      def supported_extension?(path)
+        path.end_with?(*SUPPORTED_EXTENSIONS)
+      end
+
+      # Checks if a filename is named in correspondence to the translations it loaded.
+      # The locale extracted from the path must be the single locale loaded in the translations.
+      def file_named_correctly?(path, translations)
+        locales = translations.keys.map(&:to_sym)
+        return false unless locales.one?
+
+        LocaleExtractor.locale_from_path(path) == locales.first
+      end
```

Rather than return `true/false` this method should raise a final error. This way we can include in the error message which assumption was broken, and in which file, making it much easier to fix.

---

I was on the fence about which way to go. I agree that knowing which assumption was broken will help, and this is something I'll add with the current approach.

The reason why I decided to raise with all the offending files at once is mainly a DX concern. If someone's onboarding using this strategy and would like to acquire a complete list of offending files, it would be handy to raise this in a single error.

My worry is that people will fight with repeated errors being raised and only able to fix them on a one-by-one basis.

Thoughts?

---

You are right, collecting all the offenses and raising once is even better. But I think you can do both, it just can't be expressed by simple booleans since you have multiple types of offense (e.g. path not matching at all, and content not respecting what the path claims).

So maybe you can collect a list of error message in an array and then raise if it's not empty.

---

We reached the same conclusion when pairing on this yesterday! 😄  I believe Paarth has submitted the changes upstream now: https://github.com/ruby-i18n/i18n/pull/612/files#diff-8e52202d93e7810223b33ab64de65635e73f04ffa8d7d38b14b5e656039d3ea1R152-R162